### PR TITLE
Drop checkFinalTx and use Median Time Past to check finality of wallet transactions

### PR DESCRIPTION
**Pull request by ariard**
Replace `Chain::checkFinalTx` by `CheckFinalWalletTx` by caching best block median time past which is following BIP113 rule than time-locked transactions have nLockTime set to less than the median time of the previous block they're contained in.

Fix misuage of CheckFinalTx, which was called with default flags
argument by implementation of Chain::checkFinal, triggering
finality evaluation based on GetAdjustedTime instead of consensus
and standard rules of BIP 113.

I think it should have been set at same time than d1c3762, to align wallet checks on the mempool ones.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -192,12 +192,11 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
...
 
-    // By convention a negative value for flags indicates that the
-    // current network-enforced consensus rules should be used. In
-    // a future soft-fork scenario that would mean checking which
-    // rules would be enforced for the next block and setting the
-    // appropriate flags. At the present time no soft-forks are
-    // scheduled, so no flags are set.
+    // By convention a negative value for flags indicates that no
+    // soft-forked consensus rules should be used. Setting appropriate
+    // flags select which rules would be enforced for the next block.
```
Comment (jonatack) [inline src/validation.cpp]: Here it seems should be either "Setting-appropriate flags select" or "Setting appropriate flags selects". 

I think it's the latter but it's not clear from the sentence.

Suggest: s/would/will/

File: src/validation.cpp [hunk 2]
```diff
@@ -192,12 +192,11 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
...
+    // At the present time, only CSV softfork (i.e BIP 113) alter
```
Comment (jonatack) [inline src/validation.cpp]: s/i.e/i.e./

It looks like "softfork" should be plural.

**Review by jonatack — COMMENTED** [review]
Apologies for beginning with nits rather than concept review as I lack the time today to review this properly. A few grammar suggestions for clarity, if the fundamental changes are correct.

**Review by ryanofsky — APPROVED** [review]
Code review ACK c9609737208a238984717d16e13f770ea350bb95. This seems like a good change because it should make `checkFinalTx` more accurate given current consensus rules, and also because it should make `IsTrusted` more conservative. Two suggestions for followup here or a later PR:

- It could be good to add a new test for `CWallet::IsTrusted` setting a mock time and covering this edge case. I think this should not be too hard to implement as either a c++ or a python test

- Especially with #16426, I wonder if it would be better to delete the `interfaces::Chain::checkFinalTx` method, and instead add a `CWallet::CheckFinalTx` method calling [`IsFinalTx`](https://github.com/bitcoin/bitcoin/blob/50591f6ec61b802cf4193cdbefcc85ad75716e8d/src/consensus/tx_verify.cpp#L17-L28) with `m_last_block_processed` height and time, so the wallet uses a consistent tip and there aren't race conditions when the tip is changing. I previously suggested this in https://github.com/bitcoin/bitcoin/pull/16426#discussion_r342719832). This could also be added as another Chain interface [TODO item](https://github.com/bitcoin/bitcoin/blob/988eaf2fcb71284c5fc57c7ef755bdafa4b929a0/src/interfaces/chain.h#L33-L54) if it is not worth the effort to implement now.

**Comment by ariard** [comment]
@ryanofsky Thanks for review!

I went ahead on replaced `checkFinalTx` by `CheckFinalWalletTx` which itself calls `IsFinalTx` by caching median time past in the wallet.

About testing, I think a bit about it, it's quite hard because contrary to the consensus code where finality enforcement is going to change after softfork activation, wallet is always using current finality consensus rules. So that would need to have 2 versions of the wallet. I think that's an issue already known of our current wallet test framework..

File: src/validation.cpp [hunk 3]
```diff
@@ -192,12 +192,11 @@ bool CheckFinalTx(const CTransaction &tx, int flags)
...
     AssertLockHeld(cs_main);
 
-    // By convention a negative value for flags indicates that the
-    // current network-enforced consensus rules should be used. In
-    // a future soft-fork scenario that would mean checking which
-    // rules would be enforced for the next block and setting the
-    // appropriate flags. At the present time no soft-forks are
-    // scheduled, so no flags are set.
+    // By convention a negative value for flags indicates that no
+    // soft-forked consensus rules should be used. Setting appropriate
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/validation.cpp]: I think the comment was written when BIP113 code was merged but no yet activated and scheduled for activation. It's right we don't have soft-forks scheduled now, but it's false there is no flag to set (LOCKTIME_MEDIAN_TIME_PAST)

**Comment by ariard** [comment]
01377a5

Moved `IsFinalTx` from `consensus/tx_verify.h` to `consensus/tx_check.h` as it's context-independent and so let wallet code called it without linking against server one. Should fix travis build failure.

File: src/validation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanofsky) [reply src/validation.cpp]: re: https://github.com/bitcoin/bitcoin/pull/17443#discussion_r359105468

I don't understand "it's false there is no flag to set (LOCKTIME_MEDIAN_TIME_PAST)." since this code isn't setting LOCKTIME_MEDIAN_TIME_PAST, it's setting 0...

But more broadly, I don't understand why there would be a reason to keep this code if it isn't going to serve a function in a future soft fork. Why not simplify by making the CheckFinalTx `flags` argument required instead of optional, and dropping `std::max`. and dropping this comment?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3964,6 +3965,27 @@ bool CWalletTx::IsImmatureCoinBase() const
...
+    // current chain tip (BIP 113). In the future, if there
+    // is forks changing this, wallet should be updated.
+
+    // CheckFinalWalletTx() uses m_last_block_processed_height +1
+    // to evaluate nLockTime because when IsFinalTx() is called within
+    // CBlock::AcceptBlock(), the height of the block *being*
+    // evaluated is what is used. Thus if we want to know if a
+    // transaction can be part of the *next* block, we need to call
+    // IsFinalTx() with one more than m_last_block_processed_height.
+    return IsFinalTx(tx, m_last_block_processed_height + 1, m_last_block_median_time_past);
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp]: In commit "Replace Chain::checkFinal by CWallet::CheckFinalWalletTx" (0ef43cef24b837367a90c97a31470e0f19f18552)

I see this comment is copied from [existing code](https://github.com/bitcoin/bitcoin/blob/6fef85bfa3cd7f76e83b8b57f9e4acd63eb664ec/src/validation.cpp#L200-L205), but the part about this function being called from `CBlock::AcceptBlock` doesn't apply here, right? Would it be ok to drop this long comment and simply write that we are passing  `m_last_block_processed_height + 1` because we are interested to know if the transaction could be included in the next block, not the current tip?

**Review by ryanofsky — APPROVED** [review]
Code review ACK 01377a55545beb03c6c0033f4bea483338855880 modulo some new travis failures in miner_tests.cpp. Various suggested changes since last review (thanks!)

File: src/validation.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/validation.cpp]: CheckFinalTx is called with STANDARD_LOCKTIME_VERIFY_FLAGS in `PreChecks` and `removeForReorg`. Given it's only mempool I think you'e right, you can hardcode that you always want to use the most recent consensus rules but I would prefer to do that as a follow-up to avoid missing something.

File: src/wallet/wallet.cpp (hunk 4 above)
Comment (ariard) [reply src/wallet/wallet.cpp]: The comment is about that we want to call `IsFinalTx` the same way it's called in `AcceptBlock`, i.e with `nBlockHeight` set to tip+1 to evaluate the consensus rule identically in both wallet and node. You don't want discrepancies between them. I think the last sentence is expressing that "if a transaction can be part of the *next* block", and speaking of `m_last_block_processed_height` is verbose but at least you avoid confusion between wallet tip and node tip.

**Review by ariard — COMMENTED** [review]
Thanks for review @ryanofsky! Updated at 6fb938c with fixing travis build failure.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 6fb938c78d1fed9e460775f772e276b473075965. No change since last review other than fixing travis include error. There is still a remaining travis error though:

https://travis-ci.org/bitcoin/bitcoin/jobs/626988480#L3179

```
wallet/wallet.cpp:3999:26: error: reading variable 'm_last_block_processed_height' requires holding mutex 'cs_wallet' [-Werror,-Wthread-safety-analysis]
    return IsFinalTx(tx, m_last_block_processed_height + 1, m_last_block_median_time_past);
```

**Comment by ariard** [comment]
Updated ec73698, extended both `IsTrusted` declarations with exclusive wallet lock holding macros to pass static analysis on macOS and so remove travis failure.

File: src/wallet/wallet.h [hunk 5]
```diff
@@ -477,8 +477,8 @@ class CWalletTx
     bool IsEquivalentTo(const CWalletTx& tx) const;
 
     bool InMempool() const;
-    bool IsTrusted(interfaces::Chain::Lock& locked_chain) const;
-    bool IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const;
+    bool IsTrusted(interfaces::Chain::Lock& locked_chain) const NO_THREAD_SAFETY_ANALYSIS;
+    bool IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const NO_THREAD_SAFETY_ANALYSIS;
```
Comment (ryanofsky) [inline src/wallet/wallet.h]: In commit "Replace Chain::checkFinal by CWallet::CheckFinalWalletTx" (1f479751b2eaf6f17ae76334b72ebc5fd76d8956)

I think it'd be safer to just add `AssertLockHeld` statements if the compiler can't determine that a particular lock is held, than to disable safety analysis altogether. On my system, the following changes seem to work:

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 9d291f0b585..c702f90681d 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1882,6 +1882,7 @@ bool CWalletTx::IsTrusted(interfaces::Chain::Lock& locked_chain) const
 bool CWalletTx::IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const
 {
     // Quick answer in most cases
+    AssertLockHeld(pwallet->cs_wallet);
     if (!pwallet->CheckFinalWalletTx(*tx)) return false;
     int nDepth = GetDepthInMainChain();
     if (nDepth >= 1) return true;
@@ -3967,6 +3968,7 @@ bool CWalletTx::IsImmatureCoinBase() const
 
 bool CWalletTx::isFinal() const
 {
+    AssertLockHeld(pwallet->cs_wallet);
     return pwallet->CheckFinalWalletTx(*tx);
 }
 
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index acf0de45f1c..4ae39ba328a 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -477,8 +477,8 @@ public:
     bool IsEquivalentTo(const CWalletTx& tx) const;
 
     bool InMempool() const;
-    bool IsTrusted(interfaces::Chain::Lock& locked_chain) const NO_THREAD_SAFETY_ANALYSIS;
-    bool IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const NO_THREAD_SAFETY_ANALYSIS;
+    bool IsTrusted(interfaces::Chain::Lock& locked_chain) const;
+    bool IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const;
 
     int64_t GetTxTime() const;
 
@@ -531,7 +531,7 @@ public:
     const uint256& GetHash() const { return tx->GetHash(); }
     bool IsCoinBase() const { return tx->IsCoinBase(); }
     bool IsImmatureCoinBase() const;
-    bool isFinal() const NO_THREAD_SAFETY_ANALYSIS;
+    bool isFinal() const;
 };
 
 class COutput
```



6e10c40e9507c2ff45d47ca863778fff85584b36

**Review by ryanofsky — APPROVED** [review]
Conditional code review ACK fa5b990647baf7184d1fd610f9a747be3ed46127 if the NO_THREAD_SAFETY_ANALYSIS annotations can be removed, or at least documented (see comment below). Only changes since the last review are adding the annotations and fixing up some comments in the third commit message.

File: src/wallet/wallet.h (hunk 5 above)
Comment (ariard) [reply src/wallet/wallet.h]: Did you test on a Linux ? IIRC clang static analysis only work on MacOS. `AssertLockHeld` is lock checking at runtime

**Review by ariard — COMMENTED** [review]
Updated at 9afb8dd, thanks to your suggestions @ryanofsky, `AssertLockHeld` being disabled with default flags it's not going to be a performance hit, but will hit deadlock in Travis.

File: src/wallet/wallet.h (hunk 5 above)
Comment (ryanofsky) [reply src/wallet/wallet.h]: > Did you test on a Linux ? IIRC clang static analysis only work on MacOS. `AssertLockHeld` is lock checking at runtime

AssertLockHeld does affect static analysis, because it sets the `assert_exclusive_lock` attribute, see https://clang.llvm.org/docs/ThreadSafetyAnalysis.html. I am testing on linux, and I am using clang and have static analysis enabled. The following change works fine for me with the asserts added, so still think NO_THREAD_SAFETY_ANALYSIS annotations are unneeded.

```
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -481,12 +481,12 @@ public:
     // annotation "EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)". The
     // annotation "NO_THREAD_SAFETY_ANALYSIS" was temporarily added to avoid
     // having to resolve the issue of member access into incomplete type CWallet.
-    bool IsTrusted(interfaces::Chain::Lock& locked_chain) const NO_THREAD_SAFETY_ANALYSIS;
+    bool IsTrusted(interfaces::Chain::Lock& locked_chain) const;
     // TODO: Remove "NO_THREAD_SAFETY_ANALYSIS" and replace it with the correct
     // annotation "EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)". The
     // annotation "NO_THREAD_SAFETY_ANALYSIS" was temporarily added to avoid
     // having to resolve the issue of member access into incomplete type CWallet.
-    bool IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const NO_THREAD_SAFETY_ANALYSIS;
+    bool IsTrusted(interfaces::Chain::Lock& locked_chain, std::set<uint256>& trusted_parents) const;
 
     int64_t GetTxTime() const;
 
@@ -543,7 +543,7 @@ public:
     // annotation "EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)". The
     // annotation "NO_THREAD_SAFETY_ANALYSIS" was temporarily added to avoid
     // having to resolve the issue of member access into incomplete type CWallet.
-    bool isFinal() const NO_THREAD_SAFETY_ANALYSIS;
+    bool isFinal() const;
 };
```

**Review by ryanofsky — APPROVED** [review]
Code review ACK 9afb8ddd1d9406a623d209bc5a4a5e955f0b806c. Only change since last review is to lock assertions, annotations, and comments

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3960,6 +3967,29 @@ bool CWalletTx::IsImmatureCoinBase() const
...
+bool CWalletTx::isFinal() const
+{
+    AssertLockHeld(pwallet->cs_wallet);
+    return pwallet->CheckFinalWalletTx(*tx);
+}
+
+bool CWallet::CheckFinalWalletTx(const CTransaction& tx) const
+{
+    AssertLockHeld(cs_wallet);
+    // Wallet not doing historical validation work, it's only
```
Comment (fjahr) [inline src/wallet/wallet.cpp]: nit: some grammar improvement on this paragraph

```
    // The wallet is not doing historical validation work, it's only
    // interested in the nLocktime evaluation as enforced on
    // the current chain tip (BIP 113). If there are forks changing
    // this in the future, the wallet should be updated.
```

**Review by fjahr — APPROVED** [review]
ACK 9afb8ddd1d9406a623d209bc5a4a5e955f0b806c

Reviewed code and ran tests locally. I think it's a good change both as an incremental step for wallet/node separation and for the fix on `CheckFinalTx` usage.

Just some grammar nits in a comment block. And, if you get around to it, I also saw two typos in the commit messages:
70cc88fa8bfcb3032ee8d63120921b8448b10e8e => mediam
f50723670e104457c21ca8b214e0b46bf6d23df2 => misuage

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (ariard) [reply src/wallet/wallet.cpp]: Thanks for review, will take them if I have to update PR!

**Review by ryanofsky — APPROVED** [review]
Code review ACK ba18c2cb5c23c150e5a72c3594342bdd955ce3a7. No changes other than rebase due to conflict with #17261 (trivial, just adjacent lines)

**Comment by ariard** [comment]
@fjahr @ryanofsky it has been rebased after #16426 merged, they shouldn't have any substantial changes since your last ACK but let me know if you find any issue.

**Review by ryanofsky — APPROVED** [review]
Code review ACK b14e07b227f3830120e97f44b3d6ece55bf700c6. Just rebased since previous review

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -1041,7 +1041,7 @@ static UniValue ListReceived(const CWallet* const pwallet, const UniValue& param
     for (const std::pair<const uint256, CWalletTx>& pairWtx : pwallet->mapWallet) {
         const CWalletTx& wtx = pairWtx.second;
 
-        if (wtx.IsCoinBase() || !pwallet->chain().checkFinalTx(*wtx.tx)) {
+        if (wtx.IsCoinBase() || !wtx.isFinal()) {
```
Comment (jonatack) [inline src/wallet/rpcwallet.cpp:1044]: Nice improvement here!

**Review by jonatack — COMMENTED** [review]
Concept ACK / almost ACK b14e07b227f3830120e9 but this seems unfinished. The diff isn't large; best to not add TODOs and debt for later that can be done here, tie off the loose ends like https://github.com/bitcoin/bitcoin/pull/17443#discussion_r424696041, and take care in the writing of the code docs and commit messages. Could use a test to be confident it's not adding a regression.

File: src/wallet/wallet.h [hunk 8]
```diff
@@ -555,6 +555,11 @@ class CWalletTx
     // wrong copy.
     CWalletTx(CWalletTx const &) = delete;
     void operator=(CWalletTx const &x) = delete;
+    // TODO: Remove "NO_THREAD_SAFETY_ANALYSIS" and replace it with the correct
+    // annotation "EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)". The
+    // annotation "NO_THREAD_SAFETY_ANALYSIS" was temporarily added to avoid
+    // having to resolve the issue of member access into incomplete type CWallet.
```
[the remark this answers is not in the document]
Comment (jonatack) [reply src/wallet/wallet.h:561]: See #18981

File: src/wallet/wallet.cpp [hunk 9]
```diff
@@ -3924,12 +3926,16 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(interfaces::Chain& chain,
     }
 
     const Optional<int> tip_height = chain.getHeight();
+    int64_t mtp_time;
     if (tip_height) {
         walletInstance->m_last_block_processed = chain.getBlockHash(*tip_height);
         walletInstance->m_last_block_processed_height = *tip_height;
+        chain.findBlock(walletInstance->GetLastBlockHash(), FoundBlock().mtpTime(mtp_time));
```
[the remark this answers is not in the document]
Comment (ariard) [reply src/wallet/wallet.cpp]: Solved, note I don't think using `m_last_block_processed` or `GetBlockHash` changes something on the consistency level.

File: src/wallet/wallet.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/wallet/wallet.h]: Yes thanks for the call, in fact issue was solved since. Removed.

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (ariard) [reply src/wallet/wallet.cpp]: Thanks, took comment.

**Comment by ariard** [comment]
Thanks @ryanofsky and @jonatack for reviews, took most of your suggestion, specially with better comment for top message commit.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 808dc174b14fb7504c74a818ecaa2399258cf94e. Changes since last review: updating findBlock calls mostly as suggested (still not raising error on failure), updating commit messages, comments, thread annotations

File: src/wallet/wallet.h [hunk 10]
```diff
@@ -555,6 +555,7 @@ class CWalletTx
     // wrong copy.
     CWalletTx(CWalletTx const &) = delete;
     void operator=(CWalletTx const &x) = delete;
+    bool isFinal() const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Comment (ryanofsky) [inline src/wallet/wallet.h:558]: In commit "[wallet] Replace Chain::checkFinal by CWallet::CheckFinalWalletTx" (03870881969b12d8023716897ff674ec6ec3568e)

Travis error here:

```
In file included from interfaces/wallet.cpp:24:
./wallet/wallet.h:558:51: error: use of undeclared identifier 'cs_wallet'; did you mean 'pwallet'?
    bool isFinal() const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
                                                  ^~~~~~~~~
                                                  pwallet
./threadsafety.h:31:79: note: expanded from macro 'EXCLUSIVE_LOCKS_REQUIRED'
#define EXCLUSIVE_LOCKS_REQUIRED(...) __attribute__((exclusive_locks_required(__VA_ARGS__)))
                                                                              ^
./wallet/wallet.h:272:20: note: 'pwallet' declared here
    const CWallet* pwallet;
```

https://travis-ci.org/github/bitcoin/bitcoin/jobs/688716547#L3380

Probably needs to say `pwallet->cs_wallet` not just `cs_wallet`

I wonder @ariard if you're able to test lock annotations locally since these errors seem to come up a lot. It shouldn't be hard to enable them. I just configure with `./configure CXX=clang++ --enable-debug --enable-werror` for these

Comment (jonatack) [reply src/wallet/wallet.h:558]: Seeing the same messages while building locally, the gcc 8.3 build is fine but the clang 6.0 debug/werror build errors.

Comment (jonatack) [reply src/wallet/wallet.h:558]: Building with `pwallet->cs_wallet` halts with
```
./wallet/wallet.h:558:58: error: member access into incomplete type 'const CWallet'
    bool isFinal() const EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet);
                                                         ^
./wallet/fees.h:13:7: note: forward declaration of 'CWallet'
class CWallet;
      ^
```

**Comment by jonatack** [comment]
re-almost ACK 808dc17 the changes per `git diff b14e07b 808dc17` look good (thanks!); building locally with Clang is raising as mentioned above.

File: src/wallet/wallet.h (hunk 10 above)
Comment (ariard) [reply src/wallet/wallet.h:558]: @jonatack @ryanofsky You right I didn't the warning at first, was compiling with gcc locally. Okay in fact the issue hasn't been solved wrt to `NO_THREAD_SAFETY_ANALYSIS`. 

Isn't the way to solve it cleanly would be to get `CWalletTx` its own lock instead of requiring `CWallet` one ? I don't see a better way to solve dependency declaration..

**Comment by fjahr** [comment]
I also see the build error locally (I use clang by default). Code looks good but will wait with a full review until the fix is up.

File: src/wallet/wallet.h (hunk 10 above)
Comment (ariard) [reply src/wallet/wallet.h]: Updated dropping `isFinal` accessor on the advice of @ryanofsky

**Comment by ariard** [comment]
Rebased and updated with fix at d55ba31. Thanks for your patience!

**Review by ryanofsky — APPROVED** [review]
Code review ACK d55ba31ddd0f755db763db600413b2ce31a5466c. No significant changes since last review, just replaced wtx.isFinal calls with wallet.CheckFinalWalletTx to work around thread annotation problems

**Comment by MarcoFalke** [comment]
If there is anything relevant that I missed, I think it might be best to open a fresh pull request.

**Comment by ariard** [comment]
I agree this PR is not relevant anymore in its substantial parts.
I've looked on e30b6ea, effectively we don't need anymore `checkFinalTx` as we rely directly on `CWalletTx::InMempool()` in `AvailableCoins()` and `CachedTxIsTrusted()` to filter out the non-final transactions.
Also `28bdaa3` sounds good to me, as we only evaluate transaction finality at tip, at which BIP 113 is always active.
The first commit, moving `IsFinalTx` from `tx_verify.h` to `tx_check.h` is not relevant anymore as there is no more non-server caller of the moved function.
